add challenge for safety of primitive conversions (#217)
model-checking/verify-rust-std: doc/src/challenges/0014-convert-num.md

# Challenge 14: Safety of Primitive Conversions

- **Status:** Open

> - **Tracking Issue:** *None*

Reviewer: Knowing that this is a chicken-and-egg problem: could you please create the tracking issue right away (see https://github.com/model-checking/verify-rust-std/issues/19 or several others for an example)?

Reviewer 2: Added #220

# Challenge 14: Safety of Primitive Conversions

- **Status:** Open
- **Tracking Issue:** https://github.com/model-checking/verify-rust-std/issues/220
- **Start date:** 2024/12/15
- **End date:** 2025/2/28
- **Prize:** *TBD*

-------------------

## Goal

Verify the safety of number conversions in `core::convert::num`.

There are three classes of conversions that use unsafe code. All conversions use macros to implement traits for primitive types in bulk. You will need to add contracts into the macros so that contracts are generated for each implementation. As the conversions are all macro generated, it is probably a good idea to write your own macro for generating the harnesses.

### Success Criteria

> 

Reviewer 3: ```suggestion

#### NonZero Conversions
```

Write a type invariant for `core::num::NonZero`, then write harnesses for all `nonzero` conversions.

> Write proof contracts for each NonZero primitive conversion, listed in full below. These conversions are implemented through two macros: `impl_nonzero_int_from_nonzero_int!` and `impl_nonzero_int_try_from_nonzero_int!`. For contracts using `impl_nonzero_int_from_nonzero_int!` there should be one contract for each implemented conversion. For contracts using `impl_nonzero_int_try_from_nonzero_int!` there should be two contracts for each conversion, one with an assumption that guarantees the absence of a panic and one with no assumption that is marked by a `#[should_panic]` attribute. 

Reviewer 3: ```suggestion
Write proof contracts for each NonZero primitive conversion, listed in full below. These conversions are implemented through two macros: `impl_nonzero_int_from_nonzero_int!` and `impl_nonzero_int_try_from_nonzero_int!`. 

For each invocation of `impl_nonzero_int_from_nonzero_int!`, prove that the conversion it implements does not cause undefined behavior nor panic. For example, for `impl_nonzero_int_from_nonzero_int!(u8 => u16);`, prove that calling `NonZero<u16>::from(small: NonZero<u8>)` does not cause undefined behavior nor panic for an arbitrary `small` that satisfies the `NonZero` type invariant.

For each invocation of `impl_nonzero_int_try_from_nonzero_int!`, prove that the conversion it implements does not cause undefined behavior. For example, for `impl_nonzero_int_try_from_nonzero_int!(u16 => u8);`, prove that calling `NonZero<u8>::try_from(value: NonZero<u16>)` does not cause undefined behavior for an arbitrary `value` that satisfies the `NonZero` type invariant. Additionally, prove that if the `value` does not fit into the target type (e.g., `value` is too large to fit into a `NonZero<u8>`) that the function panics.
```

Reviewer 3: Kani is the only approved tool as of now, but we'd like to get other tools involved, so the challenges should be tool-agnostic (i.e., not mention harnesses or Kani-specific attributes).

```rust
macro_rules! impl_float_to_int {
    ($Float:ty => $($Int:ty),+) => {
        #[unstable(feature = "convert_float_to_int", issue = "67057")]
        impl private::Sealed for $Float {}
        $(
            #[unstable(feature = "convert_float_to_int", issue = "67057")]
            impl FloatToInt<$Int> for $Float {
                #[inline]
                unsafe fn to_int_unchecked(self) -> $Int {
                    // SAFETY: the safety contract must be upheld by the caller.
                    unsafe { crate::intrinsics::float_to_int_unchecked(self) }
                }
            }
        )+
    }
}
```

The safety constraints referenced in the comments are that the input value must:
- Not be NaN
- Not be infinite
- Be representable in the return type Int, after truncating off its fractional part

These constraints are given in the [documentation](https://doc.rust-lang.org/std/primitive.f32.html#method.to_int_unchecked). 
 
The intrinsic corresponds to the [fptoui](https://llvm.org/docs/LangRef.html#fptoui-to-instruction)/[fptosi](https://llvm.org/docs/LangRef.html#fptosi-to-instruction) LLVM instructions, which may be useful for reference.

> 

Reviewer 3: ```suggestion
Prove safety for each of the following conversions:
impl_float_to_int!(f16 => u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize);
impl_float_to_int!(f32 => u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize);
impl_float_to_int!(f64 => u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize);
impl_float_to_int!(f128 => u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize);

```

# Challenge 14: Safety of Primitive Conversions

- **Status:** Open

> - **Tracking Issue:** *None*

Reviewer: ```suggestion
- **Tracking Issue:** https://github.com/model-checking/verify-rust-std/issues/220
```

- **Start date:** 2024/12/15
- **End date:** 2025/2/28
- **Prize:** *TBD*

-------------------

## Goal

Verify the safety of number conversions in `core::convert::num`.

There are three classes of conversions that use unsafe code. All conversions use macros to implement traits for primitive types in bulk. You will need to add contracts into the macros so that contracts are generated for each implementation. As the conversions are all macro generated, it is probably a good idea to write your own macro for generating the harnesses.

### Success Criteria

> #### Safety for Float to Int

Reviewer 3: It would be nice to also list all of the types that implement this float to int functionality, similarly to how you have the list in the section below for conversions, so that we're explicit about which proofs we expect.

Reviewer 2: I moved the full list of expected contracts to the Success Criteria section. I also added a note about having a panicking harness for the `try_from` contracts.

Reviewer 3: Thanks -- I suggested putting the NonZero header back, since I think it was better to divide between the NonZero work and the float to int work. I meant just adding the implementations for float to int like you had for NonZero--added a suggestion below to that effect.

Reviewer 2: Thanks, sorry about the misunderstanding.

Reviewer 3: No worries!

```rust
macro_rules! impl_float_to_int {
    ($Float:ty => $($Int:ty),+) => {
        #[unstable(feature = "convert_float_to_int", issue = "67057")]
        impl private::Sealed for $Float {}
        $(
            #[unstable(feature = "convert_float_to_int", issue = "67057")]
            impl FloatToInt<$Int> for $Float {
                #[inline]
                unsafe fn to_int_unchecked(self) -> $Int {
                    // SAFETY: the safety contract must be upheld by the caller.
                    unsafe { crate::intrinsics::float_to_int_unchecked(self) }
                }
            }
        )+
    }
}
```

The safety constraints referenced in the comments are that the input value must:
- Not be NaN
- Not be infinite
- Be representable in the return type Int, after truncating off its fractional part

> These constraints are given in the [documenation](https://doc.rust-lang.org/std/primitive.f32.html#method.to_int_unchecked). 

Reviewer 3: ```suggestion
These constraints are given in the [documentation](https://doc.rust-lang.org/std/primitive.f32.html#method.to_int_unchecked). 
```
